ddtrace/llmobs/_integrations/langgraph.py

```diff
@@ -0,0 +1,174 @@
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+
+from ddtrace import tracer
+from ddtrace.ext import SpanTypes
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.llmobs._constants import INPUT_VALUE
+from ddtrace.llmobs._constants import NAME
+from ddtrace.llmobs._constants import OUTPUT_VALUE
+from ddtrace.llmobs._constants import SPAN_KIND
+from ddtrace.llmobs._constants import SPAN_LINKS
+from ddtrace.llmobs._integrations.base import BaseLLMIntegration
+from ddtrace.llmobs._integrations.utils import format_langchain_io
+from ddtrace.llmobs._utils import _get_attr
+from ddtrace.llmobs._utils import _get_llmobs_parent_id
+from ddtrace.llmobs._utils import _get_nearest_llmobs_ancestor
+from ddtrace.span import Span
+
+
+class LangGraphIntegration(BaseLLMIntegration):
+    _integration_name = "langgraph"
+    _graph_nodes_by_task_id: Dict[str, Any] = {}  # maps task_id to dictionary of name, span, and span_links
+
+    def _llmobs_set_tags(
+        self,
+        span: Span,
+        args: List[Any],
+        kwargs: Dict[str, Any],
+        response: Optional[Any] = None,
+        operation: str = "",  # oneof graph, node
+    ):
+        if not self.llmobs_enabled:
+            return
+
+        inputs = get_argument_value(args, kwargs, 0, "input")
+        config = get_argument_value(args, kwargs, 1, "config", optional=True)
+        metadata = _get_attr(config, "metadata", {})
+        instance_id = metadata.get("langgraph_checkpoint_ns", "").split(":")[-1]
+        invoked_node = self._graph_nodes_by_task_id.setdefault(instance_id, {})
+        invoked_node["span"] = {"trace_id": "{:x}".format(span.trace_id), "span_id": str(span.span_id)}
+
+        span_links = [_default_span_link(span)]
+        invoked_node_span_links = invoked_node.get("span_links")
+        if invoked_node_span_links is not None:
+            span_links = invoked_node_span_links
+        current_span_links = span._get_ctx_item(SPAN_LINKS) or []
+
+        span._set_ctx_items(
+            {
+                SPAN_KIND: "agent" if operation == "graph" else "task",
+                INPUT_VALUE: format_langchain_io(inputs),
+                OUTPUT_VALUE: format_langchain_io(response),
+                NAME: self._graph_nodes_by_task_id.get(instance_id, {}).get("name") or kwargs.get("name", span.name),
+                SPAN_LINKS: current_span_links + span_links,
+            }
+        )
+        if operation == "graph" and not _is_subgraph(span):
+            self._graph_nodes_by_task_id.clear()
+
+    def llmobs_handle_pregel_loop_tick(
+        self, finished_tasks: dict, next_tasks: dict, more_tasks: bool, is_subgraph_node: bool = False
```

Comment:
Question - do we need `more_tasks` to be passed? Can we not just check if `len(next_tasks) == 0`?

Reply:
hm yeah i don't think we technically do. i think i did it this way because i wasn't sure if for some reason there'd ever be a case where `len(next_tasks) != 0` and `more_tasks == True` (although like, why would that ever be the case??). it even looks like that in the [source](https://github.com/langchain-ai/langgraph/blob/aab6fdf3f3c5ff6f695335cc0896b32da4b1dc6e/libs/langgraph/langgraph/pregel/loop.py#L486-L488). i'll do that other check for now, i doubt they'd change this logic.

Reply:
Hmm if they expose the boolean for us then I'm fine with continuing to use that, but thanks for looking into this!